File: kotlin/src/main/kotlin/advertisements/advertisement/application/command/updateadvertisement/UpdateAdvertisementUseCase.kt
package advertisements.advertisement.application.command.updateadvertisement

import advertisements.advertisement.domain.exceptions.AdvertisementNotFoundException
import advertisements.advertisement.application.exceptions.PasswordDoesNotMatchException
import advertisements.advertisement.domain.services.AdvertisementSecurityService
import advertisements.advertisement.domain.value_object.AdvertisementId
import advertisements.advertisement.domain.value_object.Description
import advertisements.shared.value_object.Email
import advertisements.shared.value_object.Password
import advertisements.shared.value_object.UserId

class UpdateAdvertisementUseCase(
    private val advertisementRepository: advertisements.advertisement.domain.AdvertisementRepository,
    private val advertisementSecurityService: AdvertisementSecurityService,
    private val transactionManager: framework.database.TransactionManager,
) {
    fun execute(updateAdvertisementCommand: UpdateAdvertisementCommand) {
        transactionManager.beginTransaction()

        try {
            val advertisementId = AdvertisementId(updateAdvertisementCommand.id)
            val advertisement = advertisementRepository.findById(advertisementId)

            if (null === advertisement) {
                throw AdvertisementNotFoundException.withId(advertisementId.value())
            }

            advertisementSecurityService.verifyMemberUserCanManageAdvertisement(
                UserId(updateAdvertisementCommand.securityUserId),
                advertisement
            )

            if (!advertisement.password.isValidatedWith(updateAdvertisementCommand.password))
                throw PasswordDoesNotMatchException.build()

            advertisement.update(
                Description(updateAdvertisementCommand.description),
                Email(updateAdvertisementCommand.email),
                Password.fromPlainPassword(updateAdvertisementCommand.password)
            )

            advertisementRepository.save(advertisement)
            transactionManager.commit()
        } catch (e: Exception) {
            transactionManager.rollback()
            throw e
        }
    }
}

File: typescript/src/advertisements/advertisement/application/command/update-advertisement/UpdateAdvertisementUseCase.ts
import { AdvertisementRepository } from "../../../domain/AdvertisementRepository"
import { UpdateAdvertisementCommand } from "./UpdateAdvertisementCommand"
import {Password} from "../../../../shared/domain/value-object/Password";
import {Description} from "../../../domain/value-object/Description";
import {AdvertisementId} from "../../../domain/value-object/AdvertisementId";
import {InvalidPasswordException} from "../../exceptions/InvalidPasswordException";
import {AdvertisementNotFoundException} from "../../../domain/exceptions/AdvertisementNotFoundException";
import {Email} from "../../../../shared/domain/value-object/Email";
import {UserId} from "../../../../shared/domain/value-object/UserId";
import {SecurityService} from "../../../domain/services/SecurityService";
import {TransactionManager} from "../../../../../framework/database/TransactionManager";

export class UpdateAdvertisementUseCase {

  constructor(
    private advertisementRepository: AdvertisementRepository,
    private securityService: SecurityService,
    private transactionManager: TransactionManager,
  ) {

  }

  async execute(command: UpdateAdvertisementCommand): Promise<void> {
    await this.transactionManager.beginTransaction()

    try {
      const advertisementId = new AdvertisementId(command.id)
      const advertisement = await this.advertisementRepository.findById(advertisementId)

      if (!advertisement) {
        throw AdvertisementNotFoundException.withId(advertisementId.value())
      }

      await this.securityService.verifyMemberUserCanManageAdvertisement(new UserId(command.securityUserId), advertisement)

      if (!await advertisement.password().isValid(command.password)) {
        throw InvalidPasswordException.build()
      }

      advertisement.update(
          new Description(command.description),
          new Email(command.email),
          await Password.fromPlainPassword(command.password)
      )

      await this.advertisementRepository.save(advertisement)
      await this.transactionManager.commit();
    } catch (error) {
      await this.transactionManager.rollback();
      throw error
    }
  }
}
